File: packages/icoder-sdk/src/composition.ts
/** Alpha-v2 multi-Agent composition primitives aligned with Corti Agent SDK. */

export interface MessageResponse {
  status?: string;
  text?: string | null;
  [key: string]: unknown;
}

export interface Runnable<I = unknown, O = MessageResponse> {
  run(input: I): Promise<O>;
}

type AnyRunnable = { run(input: any): Promise<any> };

export interface WorkflowStepConfig {
  agent: AnyRunnable;
  when?: (previous: MessageResponse) => boolean | Promise<boolean>;
  transform?: (previous: MessageResponse) => unknown | Promise<unknown>;
  retries?: number;
  retryDelay?: number;
}

export type WorkflowStep = AnyRunnable | WorkflowStepConfig;

export interface WorkflowResult {
  output: MessageResponse;
  steps: MessageResponse[];
  stoppedEarly: boolean;
}

export interface ParallelStepConfig {
  agent: AnyRunnable;
  input?: unknown;
}

export type ParallelStep = AnyRunnable | ParallelStepConfig;

export type ParallelSettledResult =
  | { status: 'fulfilled'; value: MessageResponse }
  | { status: 'rejected'; reason: unknown };

export interface ParallelResult {
  results: ParallelSettledResult[];
  fulfilled: MessageResponse[];
  rejected: unknown[];
}

function isRunnable(value: unknown): value is AnyRunnable {
  return typeof value === 'object' && value !== null &&
    typeof (value as { run?: unknown }).run === 'function';
}

function responseStatus(value: unknown): string {
  if (typeof value !== 'object' || value === null) return '';
  const status = (value as { status?: unknown }).status;
  return typeof status === 'string' ? status.toLowerCase() : '';
}

function responseText(value: unknown): string {
  if (typeof value !== 'object' || value === null) return '';
  const text = (value as { text?: unknown }).text;
  return typeof text === 'string' ? text : '';
}

function sleep(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

export class Parallel {
  readonly branches: ParallelStep[];

  constructor(branches: ParallelStep[]) {
    if (!Array.isArray(branches) || branches.length === 0) {
      throw new TypeError('parallel requires at least one branch');
    }
    for (const branch of branches) {
      const runnable = isRunnable(branch)
        ? branch
        : (branch as ParallelStepConfig | null)?.agent;
      if (!isRunnable(runnable)) throw new TypeError('parallel branch must be runnable');
    }
    this.branches = [...branches];
  }

  async run(input: unknown): Promise<ParallelResult> {
    const settled = await Promise.allSettled(this.branches.map(async (branch) => {
      if (isRunnable(branch)) return branch.run(input);
      const branchInput = Object.prototype.hasOwnProperty.call(branch, 'input')
        ? branch.input
        : input;
      return branch.agent.run(branchInput);
    }));
    const results: ParallelSettledResult[] = settled.map((item) => (
      item.status === 'fulfilled'
        ? { status: 'fulfilled', value: item.value as MessageResponse }
        : { status: 'rejected', reason: item.reason }
    ));
    return {
      results,
      fulfilled: results
        .filter((item): item is { status: 'fulfilled'; value: MessageResponse } =>
          item.status === 'fulfilled')
        .map((item) => item.value),
      rejected: results
        .filter((item): item is { status: 'rejected'; reason: unknown } =>
          item.status === 'rejected')
        .map((item) => item.reason),
    };
  }
}

export function parallel(branches: ParallelStep[]): Parallel {
  return new Parallel(branches);
}

function isWorkflowConfig(step: WorkflowStep): step is WorkflowStepConfig {
  return !isRunnable(step) && typeof step === 'object' && step !== null &&
    isRunnable((step as WorkflowStepConfig).agent);
}

async function runWorkflowStep(runnable: AnyRunnable, input: unknown): Promise<MessageResponse> {
  if (runnable instanceof Parallel) {
    const result = await runnable.run(input);
    if (result.fulfilled.length === 0) {
      const error = new Error('all parallel workflow branches failed');
      (error as Error & { rejected?: unknown[] }).rejected = result.rejected;
      throw error;
    }
    return {
      status: 'completed',
      text: result.fulfilled.map(responseText).join('\n'),
      parallel: result,
    };
  }
  const response = await runnable.run(input);
  if (typeof response !== 'object' || response === null) {
    throw new TypeError('workflow runnable returned a non-object response');
  }
  return response as MessageResponse;
}

export class Workflow {
  readonly definitions: WorkflowStep[];

  constructor(steps: WorkflowStep[]) {
    if (!Array.isArray(steps) || steps.length === 0) {
      throw new TypeError('workflow requires at least one step');
    }
    if (steps.some((step) => !isRunnable(step) && !isWorkflowConfig(step))) {
      throw new TypeError('workflow step must be runnable or a step configuration');
    }
    this.definitions = [...steps];
  }

  async run(input: unknown): Promise<WorkflowResult> {
    const executed: MessageResponse[] = [];
    let previous: MessageResponse = {
      status: 'completed',
      text: typeof input === 'string' ? input : '',
      input,
    };
    let nextInput: unknown = input;

    for (let index = 0; index < this.definitions.length; index += 1) {
      const definition = this.definitions[index];
      const config = isWorkflowConfig(definition) ? definition : undefined;
      const runnable = config?.agent ?? definition as AnyRunnable;
      if (config?.when && !(await config.when(previous))) continue;
      const stepInput = config?.transform
        ? await config.transform(previous)
        : (index === 0 ? nextInput : responseText(previous));
      const retries = config?.retries ?? 0;
      const retryDelay = config?.retryDelay ?? 1000;
      if (!Number.isInteger(retries) || retries < 0) {
        throw new RangeError('workflow retries must be a non-negative integer');
      }
      if (!Number.isFinite(retryDelay) || retryDelay < 0) {
        throw new RangeError('workflow retryDelay must be non-negative');
      }

      let response: MessageResponse | undefined;
      let lastError: unknown;
      for (let attempt = 0; attempt <= retries; attempt += 1) {
        try {
          response = await runWorkflowStep(runnable, stepInput);
          lastError = undefined;
        } catch (error) {
          lastError = error;
        }
        const needsRetry = lastError !== undefined || responseStatus(response) === 'failed';
        if (!needsRetry || attempt === retries) break;
        if (retryDelay > 0) await sleep(retryDelay);
      }
      if (lastError !== undefined) throw lastError;
      if (response === undefined) throw new Error('workflow step produced no response');
      executed.push(response);
      previous = response;
      nextInput = responseText(response);
      if (responseStatus(response) === 'failed') {
        return { output: response, steps: executed, stoppedEarly: true };
      }
    }
    return { output: previous, steps: executed, stoppedEarly: false };
  }
}

export function workflow(steps: WorkflowStep[]): Workflow {
  return new Workflow(steps);
}

export const END: unique symbol = Symbol.for('@icoder/sdk/state-graph/end');
export type StateGraphEnd = typeof END;
export type StateGraphNode<S extends Record<string, unknown>> =
  (state: S) => Partial<S> | Promise<Partial<S>>;
export type StateGraphEdge<S extends Record<string, unknown>> =
  string | StateGraphEnd | ((state: S) => string | StateGraphEnd | Promise<string | StateGraphEnd>);

export interface StateGraphStep<S extends Record<string, unknown>> {
  node: string;
  delta: Partial<S>;
  state: S;
}

export interface StateGraphResult<S extends Record<string, unknown>> {
  state: S;
  steps: StateGraphStep<S>[];
  iterations: number;
  terminatedBy: 'end' | 'maxIterations' | 'noEdge';
}

export class StateGraph<S extends Record<string, unknown>> {
  private readonly nodes = new Map<string, StateGraphNode<S>>();
  private readonly edges = new Map<string, StateGraphEdge<S>>();

  addNode(name: string, node: StateGraphNode<S>): this {
    if (!name || typeof node !== 'function') throw new TypeError('stateGraph node is invalid');
    this.nodes.set(name, node);
    return this;
  }

  addEdge(from: string, edge: StateGraphEdge<S>): this {
    if (!from || !(typeof edge === 'string' || edge === END || typeof edge === 'function')) {
      throw new TypeError('stateGraph edge is invalid');
    }
    this.edges.set(from, edge);
    return this;
  }

  async run(
    start: string,
    initialState: S,
    options: { maxIterations?: number } = {},
  ): Promise<StateGraphResult<S>> {
    const maxIterations = options.maxIterations ?? 100;
    if (!Number.isInteger(maxIterations) || maxIterations < 1) {
      throw new RangeError('maxIterations must be a positive integer');
    }
    if (!this.nodes.has(start)) throw new Error(`stateGraph start node not found: ${start}`);
    let current = start;
    let state = { ...initialState } as S;
    const steps: StateGraphStep<S>[] = [];

    while (true) {
      const node = this.nodes.get(current);
      if (!node) throw new Error(`stateGraph node not found: ${current}`);
      const delta = await node({ ...state } as S);
      if (typeof delta !== 'object' || delta === null || Array.isArray(delta)) {
        throw new TypeError(`stateGraph node ${current} returned an invalid delta`);
      }
      state = { ...state, ...delta };
      steps.push({ node: current, delta: { ...delta }, state: { ...state } as S });
      const edge = this.edges.get(current);
      if (edge === undefined) {
        return { state, steps, iterations: steps.length, terminatedBy: 'noEdge' };
      }
      const next = typeof edge === 'function' ? await edge({ ...state } as S) : edge;
      if (next === END) {
        return { state, steps, iterations: steps.length, terminatedBy: 'end' };
      }
      if (steps.length >= maxIterations) {
        return { state, steps, iterations: steps.length, terminatedBy: 'maxIterations' };
      }
      if (!this.nodes.has(next)) throw new Error(`stateGraph edge target not found: ${next}`);
      current = next;
    }
  }
}

export function stateGraph<S extends Record<string, unknown>>(): StateGraph<S> {
  return new StateGraph<S>();
}

export function agentNode<S extends Record<string, unknown>>(
  agent: AnyRunnable,
  input: (state: S) => unknown,
  merge: (response: MessageResponse) => Partial<S>,
): StateGraphNode<S> {
  if (!isRunnable(agent) || typeof input !== 'function' || typeof merge !== 'function') {
    throw new TypeError('agentNode requires an Agent handle, input callback, and merge callback');
  }
  return async (state: S) => merge(await agent.run(input(state)) as MessageResponse);
}

File: packages/icoder-python/icoder_sdk/composition.py
"""Alpha-v2 multi-Agent composition primitives aligned with Corti Agent SDK."""

from __future__ import annotations

import asyncio
import inspect
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Generic, Literal, Mapping, TypeVar


S = TypeVar("S", bound=dict[str, Any])


@dataclass
class MessageResponse:
    text: str | None = None
    status: str = "completed"
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class ParallelResult:
    results: list[dict[str, Any]]
    fulfilled: list[Any]
    rejected: list[Any]


@dataclass
class WorkflowResult:
    output: Any
    steps: list[Any]
    stopped_early: bool


@dataclass
class StateGraphStep(Generic[S]):
    node: str
    delta: dict[str, Any]
    state: S


@dataclass
class StateGraphResult(Generic[S]):
    state: S
    steps: list[StateGraphStep[S]]
    iterations: int
    terminated_by: Literal["end", "maxIterations", "noEdge"]


class _End:
    def __repr__(self) -> str:
        return "END"


END = _End()


async def _await(value: Any) -> Any:
    return await value if inspect.isawaitable(value) else value


def _runnable(value: Any) -> bool:
    return callable(getattr(value, "run", None))


def _status(value: Any) -> str:
    if isinstance(value, Mapping):
        status = value.get("status")
    else:
        status = getattr(value, "status", None)
    return status.lower() if isinstance(status, str) else ""


def _text(value: Any) -> str:
    if isinstance(value, Mapping):
        text = value.get("text")
    else:
        text = getattr(value, "text", None)
    return text if isinstance(text, str) else ""


class Parallel:
    def __init__(self, branches: list[Any]) -> None:
        if not isinstance(branches, list) or not branches:
            raise TypeError("parallel requires at least one branch")
        for branch in branches:
            runnable = branch if _runnable(branch) else (
                branch.get("agent") if isinstance(branch, Mapping) else None
            )
            if not _runnable(runnable):
                raise TypeError("parallel branch must be runnable")
        self.branches = list(branches)

    async def run(self, input: Any) -> ParallelResult:
        async def invoke(branch: Any) -> Any:
            if _runnable(branch):
                return await _await(branch.run(input))
            branch_input = branch["input"] if "input" in branch else input
            return await _await(branch["agent"].run(branch_input))

        raw = await asyncio.gather(
            *(invoke(branch) for branch in self.branches),
            return_exceptions=True,
        )
        results: list[dict[str, Any]] = []
        fulfilled: list[Any] = []
        rejected: list[Any] = []
        for item in raw:
            if isinstance(item, BaseException):
                results.append({"status": "rejected", "reason": item})
                rejected.append(item)
            else:
                results.append({"status": "fulfilled", "value": item})
                fulfilled.append(item)
        return ParallelResult(results=results, fulfilled=fulfilled, rejected=rejected)


def parallel(branches: list[Any]) -> Parallel:
    return Parallel(branches)


class Workflow:
    def __init__(self, steps: list[Any]) -> None:
        if not isinstance(steps, list) or not steps:
            raise TypeError("workflow requires at least one step")
        for step in steps:
            runnable = step if _runnable(step) else (
                step.get("agent") if isinstance(step, Mapping) else None
            )
            if not _runnable(runnable):
                raise TypeError("workflow step must be runnable or a step configuration")
        self.definitions = list(steps)

    @staticmethod
    async def _run_step(runnable: Any, input: Any) -> Any:
        if isinstance(runnable, Parallel):
            result = await runnable.run(input)
            if not result.fulfilled:
                error = RuntimeError("all parallel workflow branches failed")
                setattr(error, "rejected", result.rejected)
                raise error
            return MessageResponse(
                text="\n".join(_text(item) for item in result.fulfilled),
                metadata={"parallel": result},
            )
        response = await _await(runnable.run(input))
        if response is None or not (
            isinstance(response, Mapping) or hasattr(response, "__dict__")
        ):
            raise TypeError("workflow runnable returned a non-object response")
        return response

    async def run(self, input: Any) -> WorkflowResult:
        previous: Any = MessageResponse(
            text=input if isinstance(input, str) else "",
            metadata={"input": input},
        )
        executed: list[Any] = []
        next_input = input
        for index, definition in enumerate(self.definitions):
            config = definition if isinstance(definition, Mapping) else None
            runnable = config["agent"] if config is not None else definition
            if config is not None and config.get("when") is not None:
                if not await _await(config["when"](previous)):
                    continue
            if config is not None and config.get("transform") is not None:
                step_input = await _await(config["transform"](previous))
            else:
                step_input = next_input if index == 0 else _text(previous)
            retries = int(config.get("retries", 0)) if config is not None else 0
            retry_delay = float(config.get("retry_delay", 1.0)) if config is not None else 1.0
            if retries < 0:
                raise ValueError("workflow retries must be non-negative")
            if retry_delay < 0:
                raise ValueError("workflow retry_delay must be non-negative")

            response: Any = None
            last_error: Exception | None = None
            for attempt in range(retries + 1):
                try:
                    response = await self._run_step(runnable, step_input)
                    last_error = None
                except Exception as exc:  # preserve provider exception identity
                    last_error = exc
                needs_retry = last_error is not None or _status(response) == "failed"
                if not needs_retry or attempt == retries:
                    break
                if retry_delay:
                    await asyncio.sleep(retry_delay)
            if last_error is not None:
                raise last_error
            if response is None:
                raise RuntimeError("workflow step produced no response")
            executed.append(response)
            previous = response
            next_input = _text(response)
            if _status(response) == "failed":
                return WorkflowResult(response, executed, True)
        return WorkflowResult(previous, executed, False)


def workflow(steps: list[Any]) -> Workflow:
    return Workflow(steps)


Node = Callable[[S], Mapping[str, Any] | Awaitable[Mapping[str, Any]]]
Edge = str | _End | Callable[[S], str | _End | Awaitable[str | _End]]


class StateGraph(Generic[S]):
    def __init__(self) -> None:
        self.nodes: dict[str, Node[S]] = {}
        self.edges: dict[str, Edge[S]] = {}

    def add_node(self, name: str, node: Node[S]) -> "StateGraph[S]":
        if not name or not callable(node):
            raise TypeError("stateGraph node is invalid")
        self.nodes[name] = node
        return self

    def add_edge(self, source: str, edge: Edge[S]) -> "StateGraph[S]":
        if not source or not (isinstance(edge, str) or edge is END or callable(edge)):
            raise TypeError("stateGraph edge is invalid")
        self.edges[source] = edge
        return self

    async def run(
        self,
        start: str,
        initial_state: S,
        *,
        max_iterations: int = 100,
    ) -> StateGraphResult[S]:
        if not isinstance(max_iterations, int) or max_iterations < 1:
            raise ValueError("max_iterations must be a positive integer")
        if start not in self.nodes:
            raise KeyError(f"stateGraph start node not found: {start}")
        current = start
        state = dict(initial_state)
        steps: list[StateGraphStep[S]] = []
        while True:
            node = self.nodes.get(current)
            if node is None:
                raise KeyError(f"stateGraph node not found: {current}")
            delta = await _await(node(dict(state)))
            if not isinstance(delta, Mapping):
                raise TypeError(f"stateGraph node {current} returned an invalid delta")
            delta_dict = dict(delta)
            state.update(delta_dict)
            steps.append(StateGraphStep(
                node=current,
                delta=delta_dict,
                state=dict(state),  # type: ignore[arg-type]
            ))
            if current not in self.edges:
                return StateGraphResult(
                    state=state, steps=steps, iterations=len(steps),
                    terminated_by="noEdge",
                )
            edge = self.edges[current]
            next_node = await _await(edge(dict(state))) if callable(edge) else edge
            if next_node is END:
                return StateGraphResult(
                    state=state, steps=steps, iterations=len(steps),
                    terminated_by="end",
                )
            if len(steps) >= max_iterations:
                return StateGraphResult(
                    state=state, steps=steps, iterations=len(steps),
                    terminated_by="maxIterations",
                )
            if next_node not in self.nodes:
                raise KeyError(f"stateGraph edge target not found: {next_node}")
            current = next_node


def stateGraph() -> StateGraph[Any]:
    return StateGraph()


def agent_node(
    agent: Any,
    input_fn: Callable[[S], Any],
    merge_fn: Callable[[Any], Mapping[str, Any]],
) -> Node[S]:
    if not _runnable(agent) or not callable(input_fn) or not callable(merge_fn):
        raise TypeError("agent_node requires an Agent handle, input_fn, and merge_fn")

    async def node(state: S) -> Mapping[str, Any]:
        response = await _await(agent.run(input_fn(state)))
        return merge_fn(response)

    return node


__all__ = [
    "END", "MessageResponse", "Parallel", "ParallelResult", "StateGraph",
    "StateGraphResult", "StateGraphStep", "Workflow", "WorkflowResult",
    "agent_node", "parallel", "stateGraph", "workflow",
]
